x/wasm-storage/keeper/abci.go

```diff
@@ -70,83 +84,45 @@ func (k Keeper) ExecuteTally(ctx sdk.Context) error {
 
 	k.Logger(ctx).Info("non-empty tally list - starting tally process")
 
-	// Loop through the list to apply filter, execute tally, and post
-	// execution result.
 	var tallyList []Request
 	err = json.Unmarshal(queryRes, &tallyList)
 	if err != nil {
 		return err
 	}
-	for id, req := range tallyList {
-		filter, err := base64.StdEncoding.DecodeString(req.ConsensusFilter)
-		if err != nil {
-			return fmt.Errorf("failed to decode consensus filter: %w", err)
-		}
 
-		// Sort reveals.
-		keys := make([]string, len(req.Reveals))
-		i := 0
-		for k := range req.Reveals {
-			keys[i] = k
-			i++
-		}
-		sort.Strings(keys)
-		reveals := make([]RevealBody, len(req.Reveals))
-		for i, k := range keys {
-			reveals[i] = req.Reveals[k]
-		}
-
-		outliers, consensus, err := ApplyFilter(filter, reveals)
-		if err != nil {
-			return err
-		}
-
-		tallyID, err := hex.DecodeString(req.TallyBinaryID)
-		if err != nil {
-			return fmt.Errorf("failed to decode tally ID to hex: %w", err)
-		}
-		tallyWasm, err := k.DataRequestWasm.Get(ctx, tallyID)
-		if err != nil {
-			return fmt.Errorf("failed to get tally wasm for DR ID %d: %w", id, err)
-		}
-		tallyInputs, err := base64.StdEncoding.DecodeString(req.TallyInputs)
-		if err != nil {
-			return fmt.Errorf("failed to decode tally inputs: %w", err)
-		}
-
-		args, err := tallyVMArg(tallyInputs, reveals, outliers)
-		if err != nil {
-			return err
-		}
-
-		k.Logger(ctx).Info(
-			"executing tally VM",
-			"request_id", req.ID,
-			"tally_wasm_hash", req.TallyBinaryID,
-			"consensus", consensus,
-			"arguments", args,
-		)
-		vmRes := tallyvm.ExecuteTallyVm(tallyWasm.Bytecode, args, map[string]string{
-			"VM_MODE":   "tally",
-			"CONSENSUS": fmt.Sprintf("%v", consensus),
-		})
-
-		// Post results to the SEDA contract.
+	// Loop through the list to apply filter, execute tally, and post
+	// execution result.
+	for _, req := range tallyList {
+		// Construct sudo message to be posted to the contract and
+		// populate the results fields after filterAndTally.
 		sudoMsg := Sudo{
 			ID: req.ID,
 			Result: DataResult{
 				Version:        req.Version,
 				ID:             req.ID,
 				BlockHeight:    uint64(ctx.BlockHeight()),
-				ExitCode:       byte(vmRes.ExitInfo.ExitCode),
 				GasUsed:        "0", // TODO
-				Result:         vmRes.Result,
 				PaybackAddress: req.PaybackAddress,
 				SedaPayload:    req.SedaPayload,
-				Consensus:      consensus,
 			},
-			ExitCode: byte(vmRes.ExitInfo.ExitCode),
 		}
+
+		vmRes, consensus, err := k.filterAndTally(ctx, req)
+		if err != nil {
+			// Return module error message with exit code 255 to signify
+			// that the tally VM was not executed.
+			sudoMsg.ExitCode = 0xff
+			sudoMsg.Result.ExitCode = 0xff
```

Comment:
Is there any case where 
`sudoMsg.ExitCode` and `sudoMsg.Result.ExitCode` will be different?
**If there is no such case**, why duplicate fields?

Reply:
Just to understand better, so why we are keeping this `sudoMsg.Result.ExitCode` field?
Isn't it redundant?

Reply:
Right now they're the same, but we want to take a little more time in the future to discuss with Mario how to best handle this field. There's something to be said for keeping `ExitCode` a module level field where for example `255` signals filter failure, and `123` that the VM returned an error, and `Result.ExitCode` is exclusively for the detailed VM error.